universal-application-tool-0.0.1/app/repository/DbRepository.java

```diff
@@ -0,0 +1,107 @@
+import javax.inject.Inject;
+import models.Applicant;
+import models.Program;
+import models.Question;
+import play.db.ebean.EbeanConfig;
+
+public class DbRepository {
+
+  private final EbeanServer ebeanServer;
+  private final DatabaseExecutionContext executionContext;
+
+  @Inject
+  public DbRepository(EbeanConfig ebeanConfig, DatabaseExecutionContext executionContext) {
+    this.ebeanServer = Ebean.getServer(ebeanConfig.defaultServer());
+    this.executionContext = executionContext;
+  }
+
+  /** Return all programs in a set. */
+  public CompletionStage<Set<Program>> listPrograms() {
+    return supplyAsync(() -> ebeanServer.find(Program.class).findSet(), executionContext);
+  }
+
+  public CompletionStage<Optional<Program>> lookupProgram(String name, long version) {
+    return supplyAsync(
+        () ->
+            Optional.ofNullable(
+                ebeanServer
+                    .find(Program.class)
+                    .where()
+                    .eq("name", name)
+                    .eq("version", version)
+                    .findOne()),
+        executionContext);
+  }
+
+  public CompletionStage<Void> insertProgram(Program program) {
+    return supplyAsync(
+        () -> {
+          ebeanServer.insert(program);
+          return null;
+        },
+        executionContext);
+  }
+
+  public CompletionStage<Set<Applicant>> listApplicants() {
+    return supplyAsync(() -> ebeanServer.find(Applicant.class).findSet(), executionContext);
+  }
+
+  public CompletionStage<Optional<Applicant>> lookupApplicant(long id) {
+    return supplyAsync(
+        () -> Optional.ofNullable(ebeanServer.find(Applicant.class).setId(id).findOne()),
+        executionContext);
+  }
+
+  public CompletionStage<Void> insertApplicant(Applicant applicant) {
+    return supplyAsync(
+        () -> {
+          ebeanServer.insert(applicant);
+          return null;
+        },
+        executionContext);
+  }
+
+  public CompletionStage<Set<Question>> listQuestions() {
+    return supplyAsync(() -> ebeanServer.find(Question.class).findSet(), executionContext);
+  }
+
+  public CompletionStage<Optional<Question>> lookupQuestion(long id) {
+    return supplyAsync(
+        () -> Optional.ofNullable(ebeanServer.find(Question.class).setId(id).findOne()),
+        executionContext);
+  }
+
+  public CompletionStage<Optional<Question>> lookupQuestion(String target) {
+    return supplyAsync(
+        () ->
+            Optional.ofNullable(
+                ebeanServer
+                    .find(Question.class)
+                    .where()
+                    .jsonEqualTo("object", "target", target)
+                    .findOne()),
+        executionContext);
+  }
+
+  public CompletionStage<Void> insertQuestion(Question question) {
+    return supplyAsync(
+        () -> {
+          ebeanServer.insert(question);
+          return null;
+        },
+        executionContext);
+  }
+
+  public void flush() {
```

Comment:
Does this generally need to be called after `insert` transactions? Read [the docs](https://ebean.io/apidoc/11/io/ebean/Database.html#flush--) but it's still kind of unclear. I see that it's at least useful in tests though.

Reply:
No longer needed since `target` is not present any longer, but I also am not sure - it _is_ needed to make the test pass, though, the failed insert does not fail until you flush.